### PR TITLE
Add OIDC Tomcat Agent for Java

### PR DESCRIPTION
**Pull request by chamathns**
## Purpose
Add new module, **io.asgardio.tomcat.oidc.agent**
Add sample app module, **io.asgardio.tomcat.oidc.sample**

### CODE REVIEW

File: io.asgardio.tomcat.oidc.agent/pom.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  ~ under the License.
+  -->
+
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>tomcat-oidc-agent</artifactId>
+        <groupId>io.asgardio.tomcat.oidc.agent</groupId>
+        <version>0.1.1-SNAPSHOT</version>
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/pom.xml]: Need to start with 0.1.0-SNAPSHOT

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ * @since 0.1.1
+ */
+public class JKSLoader implements ServletContextListener {
+
+    private static final Logger logger = LogManager.getLogger(JKSLoader.class);
+
+    @Override
+    public void contextInitialized(ServletContextEvent servletContextEvent) {
+
+        // First find jks properties
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Comments must end with a period

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            } else {
+                throw new SSOAgentClientException(SSOAgentConstants.JKS_PROPERTY_FILE_PARAMETER_NAME
+                        + " context-param is not specified in the web.xml");
+            }
+
+            if (jksInputStream == null) {
+                return;
+            }
+
+            // Load properties
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Comments must end with a period

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            final Properties jksProperties = new Properties();
+            jksProperties.load(jksInputStream);
+
+            // Find and set JKS required for IS server communication
+            final URL resource =
+                    this.getClass().getClassLoader()
+                            .getResource(jksProperties.getProperty(SSOAgentConstants.KEYSTORE_NAME));
+            //use constant
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Formatting issue

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+            if (jksInputStream == null) {
+                return;
+            }
+
+            // Load properties
+            final Properties jksProperties = new Properties();
+            jksProperties.load(jksInputStream);
+
+            // Find and set JKS required for IS server communication
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Comments must end with a period

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+        return properties;
+    }
+
+    public static String getPropertyByKey(final String key) {
+
+        return properties.getProperty(key);
+    }
+
+    public void contextInitialized(ServletContextEvent servletContextEvent) {
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java:68]: Add @Overrides annotation

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            } else {
+                throw new SSOAgentClientException(SSOAgentConstants.APP_PROPERTY_FILE_PARAMETER_NAME
+                        + " context-param is not specified in the web.xml");
+            }
+        } catch (SSOAgentClientException e) {
+            logger.log(Level.FATAL, "Error while loading properties.", e);
+        }
+    }
+
+    public void contextDestroyed(ServletContextEvent servletContextEvent) {
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java:93]: Add @Overrides annotation

File: io.asgardio.tomcat.oidc.sample/pom.xml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  ~ under the License.
+  -->
+
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>tomcat-oidc-agent</artifactId>
+        <groupId>io.asgardio.tomcat.oidc.agent</groupId>
+        <version>0.1.1-SNAPSHOT</version>
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.sample/pom.xml]: Need to start with 0.1.0-SNAPSHOT

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                An error has occurred!
+            </h3>
+        </div>
+        <div class="element-padding">
+            <%=exception.getMessage()%>
+        </div>
+    </div>
+</main>
+</body>
+</html>
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp]: Add a newline

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    
+    String name = null;
+    Map<String, Object> customClaimValueMap = new HashMap<>();
+    
+    if (idToken != null) {
+        try {
+            final User user = sessionContext.getUser();
+            customClaimValueMap = user.getAttributes();
+            name = user.getSubject();
+        } catch (Exception e) {
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp]: Can we catch the specific exception?

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            e.printStackTrace();
```
Comment (gayashanbc) [inline io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp]: Can we use a logger instead?

File: pom.xml [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    </parent>
+
+    <groupId>io.asgardio.tomcat.oidc.agent</groupId>
+    <modules>
+        <module>io.asgardio.tomcat.oidc.agent</module>
+        <module>io.asgardio.tomcat.oidc.sample</module>
+    </modules>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>tomcat-oidc-agent</artifactId>
+    <version>0.1.1-SNAPSHOT</version>
```
Comment (gayashanbc) [inline pom.xml]: Need to start with 0.1.0-SNAPSHOT

Comment (chamathns) [reply pom.xml]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp (hunk 11 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp (hunk 10 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp (hunk 9 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.sample/pom.xml (hunk 8 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.sample/pom.xml]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java (hunk 7 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java:93]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java (hunk 6 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/SSOAgentContextEventListener.java:68]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java (hunk 5 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java (hunk 4 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java (hunk 3 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java (hunk 2 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/src/main/java/io/asgardio/tomcat/oidc/agent/JKSLoader.java]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

File: io.asgardio.tomcat.oidc.agent/pom.xml (hunk 1 above)
Comment (chamathns) [reply io.asgardio.tomcat.oidc.agent/pom.xml]: Resolved with https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/1/commits/6b6ec648bf4aa46847e8c22e02fe4500e8a7d18e.

**Review by gayashanbc — APPROVED** [review]

File: io.asgardio.tomcat.oidc.sample/src/main/resources/oidc-sample-app.properties [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+consumerKey=KE4OYeY_gfYwzQbJa9tGhj1hZJMa
+consumerSecret=_ebDU3prFV99JYgtbnknB0z0dXoa
```
Comment (darshanasbg) [inline io.asgardio.tomcat.oidc.sample/src/main/resources/oidc-sample-app.properties:19]: Let remove values like these

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp [hunk 14]
```diff
@@ -0,0 +1,68 @@
+<%@ page import="io.asgardio.java.oidc.sdk.exception.SSOAgentException" %>
```
Comment (darshanasbg) [inline io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp:1]: Move imports after the licenses.

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp [hunk 15]
```diff
@@ -0,0 +1,117 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (darshanasbg) [inline io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp:1]: Is this required?

**Review by darshanasbg — APPROVED** [review]

File: io.asgardio.tomcat.oidc.sample/src/main/resources/oidc-sample-app.properties (hunk 13 above)
Comment (gayashanbc) [reply io.asgardio.tomcat.oidc.sample/src/main/resources/oidc-sample-app.properties:19]: Resolved in https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/7/commits/bf4f7fa7367a16bd78c9cbabc954df5fb7451e18

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp (hunk 14 above)
Comment (gayashanbc) [reply io.asgardio.tomcat.oidc.sample/src/main/webapp/error.jsp:1]: Resolved in https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/7/commits/bf4f7fa7367a16bd78c9cbabc954df5fb7451e18

File: io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp (hunk 15 above)
Comment (gayashanbc) [reply io.asgardio.tomcat.oidc.sample/src/main/webapp/home.jsp:1]: Resolved in https://github.com/asgardio/asgardio-tomcat-oidc-agent/pull/7/commits/bf4f7fa7367a16bd78c9cbabc954df5fb7451e18

**Review by darshanasbg — APPROVED** [review]

**Pull request merged**